File: FormacaoJava/DesginPatterns/src/br/com/alura/curso1/Conta.java
package br.com.alura.curso1;

public class Conta {
	private String titular;
	private double saldo;
	
	public Conta(String titular, double saldo) {
		this.setTitular(titular);
		this.setSaldo(saldo);		
	}

	public String getTitular() {
		return titular;
	}

	public void setTitular(String titular) {
		this.titular = titular;
	}

	public double getSaldo() {
		return saldo;
	}

	public void setSaldo(double saldo) {
		this.saldo = saldo;
	}
}

File: FormacaoKotlin/curso/src/br/com/alura/bytebank/models/Conta.kt
package br.com.alura.bytebank.models

import br.com.alura.bytebank.exceptions.SaldoInsuficienteException

abstract class Conta(val titular: Cliente, val numero: Int, saldo: Double) {

    var saldo = saldo
        protected set


    /**
     * Análogo a uma variável estática
     * */
    companion object Contador {
        var total: Int = 0
            private set
    }

    /**
     * Incrementa o contador global
     * */
    init {
        Conta.total++
    }

    fun deposita(valor: Double) {
        if (valor > 0) {
            this.saldo += valor
        }
    }

    open fun saca(valor: Double) {
        if (saldo < valor) {
            SaldoInsuficienteException().message
        }
        this.saldo -= valor

    }

    fun transfere(destino: Conta, valor: Double): Boolean {
        if (saldo < valor) {
            SaldoInsuficienteException().message
        }
        this.saca(valor)
        destino.deposita(valor)
        return true
    }
}